tools/rpkg/R/backend-dbplyr__duckdb_connection.R

```diff
@@ -0,0 +1,383 @@
+  sql_cast <- pkg_method("sql_cast", "dbplyr")
+  sql_paste <- pkg_method("sql_paste", "dbplyr")
+  sql_aggregate <- pkg_method("sql_aggregate", "dbplyr")
+  sql_aggregate_2 <- pkg_method("sql_aggregate_2", "dbplyr")
+  win_aggregate <- pkg_method("win_aggregate", "dbplyr")
+  win_aggregate_2 <- pkg_method("win_aggregate_2", "dbplyr")
+  win_over <- pkg_method("win_over", "dbplyr")
+  win_current_order <- pkg_method("win_current_order", "dbplyr")
+  win_current_group <- pkg_method("win_current_order", "dbplyr")
+
+
+  base_scalar <- pkg_method("base_scalar", "dbplyr")
+  base_agg <- pkg_method("base_agg", "dbplyr")
+  base_win <- pkg_method("base_win", "dbplyr")
+
+  sql_variant(
+    sql_translator(
+      .parent = base_scalar,
+      as.raw = sql_cast("VARBINARY"),
+      `%%` = function(a, b) build_sql("(CAST((", a, ") AS INTEGER)) % (CAST((", b, ") AS INTEGER))"),
+      `%/%` = function(a, b) build_sql("(((CAST((", a, ") AS INTEGER)) - (CAST((", a, ") AS INTEGER)) % (CAST((", b, ") AS INTEGER)))) / (CAST((", b, ") AS INTEGER))"),
+      `^` = sql_prefix("POW", 2),
+      bitwOr = function(a, b) sql_expr((CAST((!!a) %AS% INTEGER)) | (CAST((!!b) %AS% INTEGER))),
+      bitwAnd = function(a, b) sql_expr((CAST((!!a) %AS% INTEGER)) & (CAST((!!b) %AS% INTEGER))),
+      bitwXor = function(a, b) sql_expr(XOR((CAST((!!a) %AS% INTEGER)), (CAST((!!b) %AS% INTEGER)))),
+      bitwNot = function(a) sql_expr(~ (CAST((!!a) %AS% INTEGER))),
+      bitwShiftL = function(a, b) sql_expr((CAST((!!a) %AS% INTEGER)) %<<% (CAST((!!b) %AS% INTEGER))),
+      bitwShiftR = function(a, b) sql_expr((CAST((!!a) %AS% INTEGER)) %>>% (CAST((!!b) %AS% INTEGER))),
+      log = function(x, base = exp(1)) {
+        if (isTRUE(all.equal(base, exp(1)))) {
+          sql_expr(LN(!!x))
+        } else {
+          sql_expr(LOG(!!x) / LOG(!!base))
+        }
+      },
+      log10 = sql_prefix("LOG10", 1),
+      log2 = sql_prefix("LOG2", 1),
+
+      # Function 	    Inf 	–Inf 	NaN 	NA
+      # is.finite() 	  FALSE FALSE FALSE FALSE
+      # is.infinite() 	TRUE 	TRUE 	FALSE FALSE
+      # is.nan() 	    FALSE FALSE TRUE 	FALSE
+      # is.na() 	      FALSE FALSE TRUE 	TRUE
+      is.na = function(a) build_sql("(", a, " IS NULL OR PRINTF('%f',", a, ") = 'nan')"),
+      is.nan = function(a) build_sql("(", a, " IS NOT NULL AND PRINTF('%f',", a, ") = 'nan')"),
+      is.infinite = function(a) build_sql("(", a, " IS NOT NULL AND REGEXP_MATCHES(PRINTF('%f',", a, "),'inf'))"),
+      is.finite = function(a) build_sql("(NOT (", a, " IS NULL OR REGEXP_MATCHES(PRINTF('%f',", a, "),'inf|nan')))"),
+      grepl = duckdb_grepl,
+
+      # Return índex where first match starts, -1 if not found
+      regexpr = function(p, x) {
+        build_sql(
+          "(LENGTH(LIST_EXTRACT(STRING_SPLIT_REGEX(", x, ", ", p,
+          "),0))-(LENGTH(LIST_EXTRACT(STRING_SPLIT_REGEX(", x, ", ", p,
+          "),0))+2)/(LENGTH(", x, ")+2)*(LENGTH(", x, ")+2)+1)"
+        )
```

Comment:
Yes, I agree that this is a complicated one. The aim is to find a start of the first match, so the basic idea is rather simple: Split with regex, take the first part and check the length (and add 1 to get R/string indexing). So it is almost there with:
```sql
(LENGTH(LIST_EXTRACT(STRING_SPLIT_REGEX(txt, pattern),0))+1)
```
The only difference is that no match gives the length of string instead of -1. The rest of the formula just transforms no matches to -1. That can be achieved by first dividing by the length of string+2 (that division casts to integers, so all matches become zero as match is before the end of string) and then multiplying "back" gives zero for matches and length of string+2 for no matches. By subtracting those from the length of first match or string length gives -2 for no matches and zero indexed start point, so adding one yields the desired result. Doing that in DuckDB SQL without possibility to condition or save partial calculations turned out to be tricky, and I just used the one that could be easily achieved. It works, but I have no idea would there be simpler ways to get the same results.

If we are willing to sacrifice the -1 compatibility with R-function, then it simplifies quite nicely and only discrepancy is that no matches are at the "other" end of the string than with R-function. Probably that would be a tolerable exception from the one-to-one mapping?

In any case, I would keep some version of this in the translations. And `regexpr()` is a "better" option than `str_locate()` as the latter gives always start and end points for the first match, i.e. two dimensions in a structure instead of a single column and it would be difficult to get columns names to be the same as in R, so I would drop a support for that at the moment. 

I also noticed that DuckDB has an (undocumented) `STRPOS`-function giving the index of beginning of the match, but it doesn't support regular expressions. Otherwise that would have been useful.

Reply:
I changed this now to the shorter version that returns the `length of string +1` instead of `-1` if there is no match. So not completely consistent with R, but usable in most cases.

Reply:
Changed to simplified version that gives -1 in the case of no match (consistent with R).